library/src/main/java/com/novoda/downloadmanager/MigrationExtractor.java

```diff
@@ -35,26 +35,31 @@
             String batchTitle = batchesCursor.getString(TITLE_COLUMN);
             long downloadedDateTimeInMillis = batchesCursor.getLong(MODIFIED_TIMESTAMP_COLUMN);
 
-            Cursor downloadsCursor = database.rawQuery(DOWNLOADS_QUERY, batchId);
             Batch.Builder newBatchBuilder = new Batch.Builder(DownloadBatchIdCreator.createFrom(batchId), batchTitle);
-            List<Migration.FileMetadata> fileMetadataList = new ArrayList<>();
-
-            while (downloadsCursor.moveToNext()) {
-                String uri = downloadsCursor.getString(URI_COLUMN);
-                String originalFileName = downloadsCursor.getString(FILE_NAME_COLUMN);
-                newBatchBuilder.addFile(uri);
-
-                long rawFileSize = downloadsCursor.getLong(FILE_SIZE_COLUMN);
-                FileSize fileSize = new LiteFileSize(rawFileSize, rawFileSize);
-                Migration.FileMetadata fileMetadata = new Migration.FileMetadata(originalFileName, fileSize, uri);
-                fileMetadataList.add(fileMetadata);
-            }
-            downloadsCursor.close();
+            List<Migration.FileMetadata> fileMetadataList = extractFileMetadataFrom(batchId, newBatchBuilder);
 
             Batch batch = newBatchBuilder.build();
             migrations.add(new Migration(batch, fileMetadataList, downloadedDateTimeInMillis));
         }
         batchesCursor.close();
         return migrations;
     }
+
+    private List<Migration.FileMetadata> extractFileMetadataFrom(String batchId, Batch.Builder newBatchBuilder) {
+        Cursor downloadsCursor = database.rawQuery(DOWNLOADS_QUERY, batchId);
+        List<Migration.FileMetadata> fileMetadataList = new ArrayList<>();
+
+        while (downloadsCursor.moveToNext()) {
```

Comment:
we should protect all cursor operations with a try{}{cursor.close} ;) also review the rest code in this class regarding cursors